### PR TITLE
SE: Increase VisitCount for loop conditions to 3

### LINKED ISSUE #5601: S2583 FP: Variable change not detected inside a loop
### Description

The following method when executed returns true, but SonarAnalyser gives a warning on the last "if" statement that the expression is always false.

If you delete the 2nd last "if" the warning goes away

### Repro steps

````c#
namespace ConsoleApp1
{
    public static class Test
    {
        public static bool Run()
        {
            int pos = 0;
            bool foundA = false;
            bool readA = false;
            string test = "ab";

            while (pos < test.Length)
            {
                if (test[pos] == 'a')
                {
                    foundA = true;
                    readA = true;
                }
                else if (readA)
                {
                    readA = false;
                }
                pos++;
            }

            if (readA)
            {
                return false;
            }

            if (foundA)
            {
                return true;
            }
            return false;
        }
    }
}

````
### Expected behavior

No warnings

### Actual behavior

On the line ````if (foundA)````
Warning	S2583	Change this condition so that it does not always evaluate to 'false'; some subsequent code is never executed.	

### Known workarounds

Supress the warning

### Related information

.NET 6, Sonar Analyser.CSharp 8.38.0.46746

### PR DESCRIPTION
**Pull request by Tim-Pohlmann**
Fixing this FP:
```
var b = false;
for (var i = 0; i < 10; i++)
{
    if (i > 5)
        b = arg;
}
if (b) // FP
{
    Console.WriteLine();
}
```
The loop is explored twice. The second iteration can learn interesting constraints because `i` has `NumberConstraint.From(1, 9)`. Only visiting the loop condition twice would mean losing that information. If we visit it one more time, we can bring the information out of the loop.

There is a follow-up PR which will simplify a lot of the branching: #7797

Also:
Fixes #5601

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RoslynSymbolicExecution.cs [hunk 1]
```diff
@@ -74,7 +74,12 @@ public void Execute()
                 return;
             }
             var current = queue.Dequeue();
-            if (visited.Add(current) && current.AddVisit() <= MaxOperationVisits)
+            if (visited.Add(current)
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RoslynSymbolicExecution.cs:77]: What is the performance impact of this change?

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RoslynSymbolicExecution.cs [hunk 2]
```diff
@@ -74,7 +74,12 @@ public void Execute()
...
+                && current.AddVisit() is var visitCount
+                && (visitCount <= MaxOperationVisits
+                || (visitCount == MaxOperationVisits + 1
+                    && current.Block.ConditionalSuccessor is not null
+                    && syntaxClassifier.IsInLoopCondition(current.Block.BranchValue?.Syntax))))
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RoslynSymbolicExecution.cs:82]: Why is this needed instead of increasing the constant that is designed to control the amount?

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RoslynSymbolicExecution.cs (hunk 1 above)
Comment (Tim-Pohlmann) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RoslynSymbolicExecution.cs:77]: Two things:
- A few more branches will escape loops than before
- We call `IsInLoopCondition` a few more times
I'm not sure what that means in performance but it should be compensated by two follow-up PRs that will a) remove the "wrong" branch leaving the loop and b) remove calls to `IsInLoopCondition` during the operation handling.

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RoslynSymbolicExecution.cs (hunk 2 above)
Comment (Tim-Pohlmann) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RoslynSymbolicExecution.cs:82]: We need `MaxOperationVisits` for regular operations and `MaxOperationVisits + 1` for loop conditions.

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RoslynSymbolicExecution.cs (hunk 1 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RoslynSymbolicExecution.cs:77]: We need numbers. What happens when you analyze some of the ITs several times in average? Before vs. after

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Loops.cs [hunk 3]
```diff
@@ -38,11 +38,11 @@ public void Loops_InstructionVisitedMaxTwice(string loop)
             Tag("End", arg);
             """;
         var validator = SETestContext.CreateCS(code, "int arg, int[] items", new AddConstraintOnInvocationCheck(), new PreserveTestCheck("arg")).Validator;
-        validator.ValidateExitReachCount(2);    // PreserveTestCheck is needed for this, otherwise, variables are thrown away by LVA when going to the Exit block
-        validator.TagValues("End").Should().HaveCount(2)
-            .And.SatisfyRespectively(
-                x => x.Should().HaveOnlyConstraints(ObjectConstraint.NotNull),
-                x => x.Should().HaveOnlyConstraints(ObjectConstraint.NotNull, TestConstraint.First));
+        validator.ValidateExitReachCount(3);    // PreserveTestCheck is needed for this, otherwise, variables are thrown away by LVA when going to the Exit block
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Loops.cs:41]: UT needs a rename

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Loops.cs [hunk 4]
```diff
@@ -61,7 +61,9 @@ public void Loops_InstructionVisitedMaxTwice_For_FixedCount(string loop)
             """;
         var validator = SETestContext.CreateCS(code, "int arg", new AddConstraintOnInvocationCheck()).Validator;
         validator.ValidateExitReachCount(1);
-        validator.TagValue("End").Should().HaveOnlyConstraints(ObjectConstraint.NotNull, TestConstraint.First);   // Loop was entered, arg has only it's final constraints after looping once
+        validator.TagValues("End").Should().SatisfyRespectively(    // Loop was entered, arg has only it's final constraints after looping once or twice
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Loops.cs:64]: UT needs rename.

Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Loops.cs:64]: The comment is confusing. Is it once, or is it twice?

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Loops.cs (hunk 3 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Loops.cs:41]: On multiple places in this file

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Loops.cs [hunk 5]
```diff
@@ -257,8 +272,12 @@ public void Loops_While_FixedCount()
         validator.TagValues("Inside").Should().SatisfyRespectively(
             x => x.Should().HaveOnlyConstraints(ObjectConstraint.NotNull, NumberConstraint.From(1)),
             x => x.Should().HaveOnlyConstraints(ObjectConstraint.NotNull, NumberConstraint.From(2, 10)));
-        validator.TagValue("After").Should().HaveOnlyConstraints(ObjectConstraint.NotNull, NumberConstraint.From(10, null));
-        validator.TagValue("End").Should().HaveOnlyConstraints(ObjectConstraint.NotNull, TestConstraint.First);   // arg has only it's final constraints after looping once
+        validator.TagValues("After").Should().SatisfyRespectively(
+            x => x.Should().HaveOnlyConstraints(ObjectConstraint.NotNull, NumberConstraint.From(10, null)),
+            x => x.Should().HaveOnlyConstraints(ObjectConstraint.NotNull, NumberConstraint.From(10)));
+        validator.TagValues("End").Should().SatisfyRespectively(    // arg has only it's final constraints after looping once or twice
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Loops.cs:278]: So once? Or twice?

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/ConditionEvaluatesToConstant.CSharp8.cs [hunk 6]
```diff
@@ -392,15 +392,15 @@ void Method()
...
+            if (sb is null)     // Noncompliant FP
             {
                 sb = new StringBuilder();
             }
 
             sb.Append(i);
         }
 
-        if (sb is null) // Noncompliant FP
+        if (sb is null) // FN
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/ConditionEvaluatesToConstant.CSharp8.cs:403]: FP cannot turn into FN

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RoslynSymbolicExecution.cs (hunk 2 above)
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RoslynSymbolicExecution.cs:82]: This combo of && || && got too big for begin in the main SE loop. Let's extract it into a method `CheckVisitCount` or something like that.

It's not going to be an arrow method to make it more readable. I'd format it as
..<=..
|| (.. <= +1 && && )

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RoslynSymbolicExecution.cs [hunk 7]
```diff
@@ -74,7 +74,12 @@ public void Execute()
                 return;
             }
             var current = queue.Dequeue();
-            if (visited.Add(current) && current.AddVisit() <= MaxOperationVisits)
+            if (visited.Add(current)
+                && current.AddVisit() is var visitCount
+                && (visitCount <= MaxOperationVisits
+                || (visitCount == MaxOperationVisits + 1
+                    && current.Block.ConditionalSuccessor is not null
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RoslynSymbolicExecution.cs:81]: What's this line for? If I delete it, `RoslynSymbolicExecutionTest` is green.
```suggestion
```

**Review by pavel-mikula-sonarsource — CHANGES_REQUESTED** [review]
This PR is touching a danger zone, so we need to be careful

**Review by mary-georgiou-sonarsource — CHANGES_REQUESTED** [review]
PR looks good - brings quite some value.
As discussed Tim will update the tests to be more explicit about the way SE is visiting loops.
I'll post IT results on build time soon.

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Loops.cs (hunk 3 above)
Comment (Tim-Pohlmann) [reply analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Loops.cs:42]: I have adjusted the tests to make it easier to understand. Review per commit to see the exact changes.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/ConditionEvaluatesToConstant.CSharp8.cs (hunk 6 above)
Comment (Tim-Pohlmann) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/ConditionEvaluatesToConstant.CSharp8.cs:403]: The FP was an error. This used to be a TP (found by pure chance) which we unfortunately lost.

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RoslynSymbolicExecution.cs (hunk 7 above)
Comment (Tim-Pohlmann) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RoslynSymbolicExecution.cs:82]: It does not affect logic. It avoids more complex checks for non-conditional blocks.

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RoslynSymbolicExecution.cs (hunk 1 above)
Comment (mary-georgiou-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RoslynSymbolicExecution.cs:77]: @pavel-mikula-sonarsource  https://docs.google.com/spreadsheets/d/1sZxX2uzYnDjFTbGAZm9dKUNqkDgJj48VzadNzYq41ws/edit#gid=0

There's some increase - but IMO it's expected.

**Review by mary-georgiou-sonarsource — COMMENTED** [review]
LGTM from my side if the stats are acceptable.
Please also update the test names from `InstructionVisited*` to `LoopBodyVisited*`

Regarding the code smells for me you can ignore.

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Loops.cs (hunk 4 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Loops.cs]: The comments are adding more confusion now, because their original purpose meaning got removed. Just remove all 
`Loop was entered, arg has only its final constraints`
`loop once`
`loop twice`
and it's going to be clear enough when there's `SatisfyRespectively`

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RoslynSymbolicExecution.cs [hunk 8]
```diff
@@ -93,6 +95,17 @@ public void Execute()
...
     }
 
+    private bool CheckVisitCount(ExplodedNode current, int visitCount)
+    {
+        return visitCount <= MaxOperationVisits
+            || (visitCount <= MaxOperationVisits + 1 && IsLoopCondition(current));
+
+        bool IsLoopCondition(ExplodedNode current) =>
+            current.Block.ConditionalSuccessor is not null                              // avoid further checks for non-conditional branching blocks
+            && !current.Block.Operations.Contains(current.Operation?.Instance)          // operation is `null`, `BranchValue` or one of `BranchValue`s children
```
[the remark this answers is not in the document]
Comment (Tim-Pohlmann) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RoslynSymbolicExecution.cs:105]: I found a different approach!

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RoslynSymbolicExecution.cs (hunk 1 above)
Comment (mary-georgiou-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RoslynSymbolicExecution.cs]: @pavel-mikula-sonarsource  Updated - the numbers seem unstable + sometimes the percentage is +74% but it's worth seeing the numbers too - as it's from 1 second to 1.75 second which is not a big difference.

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RoslynSymbolicExecution.cs [hunk 9]
```diff
@@ -93,6 +93,17 @@ public void Execute()
...
     }
 
+    private bool CheckVisitCount(ExplodedNode current, int visitCount)
+    {
+        return visitCount <= MaxOperationVisits
+            || (visitCount <= MaxOperationVisits + 1 && IsLoopCondition(current));
+
+        bool IsLoopCondition(ExplodedNode current) =>
+            current.Block.ConditionalSuccessor is not null  // avoid further checks for non-conditional branching blocks
+            && (current.Operation?.Instance ?? current.Block.BranchValue) is { } branchValueOperation
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RoslynSymbolicExecution.cs:103]: I find this very cryptic. It says, "give me any operation, or branch value on branching", but what it really wants to do is "if operation is part of branching, or it's branching, check that"
```
        bool IsLoopCondition() =>
            node.Block.BranchValue is not null
            && (node.Operation is null || IsInBranchValue(node.Operation.Instance))
            && syntaxClassifier.IsInLoopCondition(node.Block.BranchValue.Syntax);  // currently processing a loop condition

        bool IsInBranchValue(IOperation current)
        {
            while (current is not null)
            {
                if (current == node.Block.BranchValue)
                {
                    return true;
                }
                current = new IOperationWrapperSonar(current).Parent;
            }
            return false;
        }
```

**Review by pavel-mikula-sonarsource — CHANGES_REQUESTED** [review]
The current logic is quite cryptic

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RoslynSymbolicExecution.cs (hunk 1 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RoslynSymbolicExecution.cs]: Any row below 30s doesn't bring value, 74% of 1sec operation is random variance.

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RoslynSymbolicExecution.cs (hunk 9 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RoslynSymbolicExecution.cs:103]: ..it took me quite some debugging and experimenting time to decode the line

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RoslynSymbolicExecution.cs (hunk 1 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RoslynSymbolicExecution.cs]: Can you update the table? To delete the useless data and have only data that tell something? I still don't understand where we stand.

**Review by pavel-mikula-sonarsource — CHANGES_REQUESTED** [review]
One last line to fix

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RoslynSymbolicExecution.cs (hunk 1 above)
Comment (mary-georgiou-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RoslynSymbolicExecution.cs]: Please check the two green columns at the bottom, they reflect the difference in the average run times between the master and the branch with the loop improvements - I deleted the percentages as they are misleading.

Negative means the loop improvements branch is faster.

Seems that the hit is not big except for two projects that we have +8 seconds - but not sure how significant this is.

Comment (Tim-Pohlmann) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RoslynSymbolicExecution.cs]: I have added some sums. I don't feel like this data is statistically sound. There are run-tim variations from over 1:40 in the same build. If we want to actually compare them, we need to run them much more often.

Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RoslynSymbolicExecution.cs]: The data still doesn't make any sense to me. What does positive `59:59.95` mean? That we are 59 minutes slower on something that took 0.1 sec before?

I suggest throwing it all away. Run akka.net from ITs 5x on master, 5x on the branch and compare the time.

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RoslynSymbolicExecution.cs [hunk 10]
```diff
@@ -93,6 +93,30 @@ public void Execute()
         checks.ExecutionCompleted();
     }
 
+    private bool CheckVisitCount(ExplodedNode node, int visitCount)
+    {
+        return visitCount <= MaxOperationVisits
+            || (visitCount <= MaxOperationVisits + 1 && IsLoopCondition());
+
+        bool IsLoopCondition() =>
+            node.Block.BranchValue is not null  // avoid further checks for blocks without BranchValue
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RoslynSymbolicExecution.cs:102]: Comment not needed anymore
```suggestion
            node.Block.BranchValue is not null
```

**Review by pavel-mikula-sonarsource — APPROVED** [review]
LGTM.
I'm still curious about the performance data, because I have no idea.

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RoslynSymbolicExecution.cs (hunk 1 above)
Comment (mary-georgiou-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RoslynSymbolicExecution.cs]: @pavel-mikula-sonarsource Apologies, I'm not familiar at all with google sheets. Data were converted to time instead of staying as duration.

Now it's fixed.

I won't delete them - now they make sense although they are indeed statistically insignificant. 

I'll do also the test with Akka and come back with the results.

Comment (mary-georgiou-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RoslynSymbolicExecution.cs]: @pavel-mikula-sonarsource  https://docs.google.com/spreadsheets/d/1sZxX2uzYnDjFTbGAZm9dKUNqkDgJj48VzadNzYq41ws/edit#gid=979078328

Here's Akka run 6 times with master and the loop improvements. 
Hope this makes sense.

Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RoslynSymbolicExecution.cs]: Ok, looks good
